File: libraries/rpc-router/python/src/rpc_router/backends/websockets/transport.py
import logging

import websockets
from websockets.protocol import State

from ...exceptions import ConnectionBrokenException, ConnectionRefusedException
from ...lifecycle import AbstractClient, AbstractServer
from ...protocol import AbstractDuplexConnection, ConnectionManager, DuplexRouter

logger = logging.getLogger(__name__)


class WebsocketsDuplexConnection(AbstractDuplexConnection):
    """Framework specific pipe translation layer handling raw string frames."""

    def __init__(
        self,
        websocket,
        router: DuplexRouter,
        client_id: str | None = None,
        manager: ConnectionManager | None = None,
    ):
        super().__init__(router, client_id=client_id, manager=manager)
        self.ws = websocket

    async def _raw_send(self, payload_str: str) -> None:
        try:
            await self.ws.send(payload_str)
        except websockets.exceptions.ConnectionClosed as e:
            raise ConnectionBrokenException from e

    async def _raw_recv(self) -> str | None:
        try:
            return await self.ws.recv()
        except websockets.exceptions.ConnectionClosed:
            return None

    async def _raw_close(self) -> None:
        await self.ws.close()


class WebsocketServer(AbstractServer):
    """Concrete WebSocket Server mapping lifecycle hooks directly to the websockets library."""

    def __init__(self, router: DuplexRouter, host: str, port: int):
        super().__init__(router)
        self.host = host
        self.port = port
        self._server = None

    async def _raw_start(self) -> None:
        """Concrete implementation bootstrapping the websockets loop server handle."""

        async def handler(websocket):
            # Context entry handles manager tracking and listener loop tasks automatically
            async with WebsocketsDuplexConnection(
                websocket, self.router, manager=self.manager
            ) as conn:
                # Abstract helper blocks the function handle until the client disconnects
                await conn.wait_forever()

        self._server = await websockets.serve(handler, self.host, self.port)
        logger.info(f"WebSocket Server active on ws://{self.host}:{self.port}")

    async def _raw_stop(self) -> None:
        """Concrete implementation shutting down the web socket network listener descriptor."""
        if self._server:
            self._server.close()
            await self._server.wait_closed()
            logger.info("WebSocket Server shut down cleanly")


class WebsocketClient(AbstractClient):
    """Concrete WebSocket Client mapping connection handshakes directly to the websockets library."""

    def __init__(self, router: DuplexRouter, target_url: str):
        super().__init__(router)
        self.target_url = target_url

    async def _create_connection(self) -> WebsocketsDuplexConnection:
        websocket = await websockets.connect(self.target_url)
        if not websocket.state == State.OPEN:
            raise ConnectionRefusedException
        return WebsocketsDuplexConnection(websocket, self.router)

File: libraries/rpc-router/ts/src/backends/websockets/transport.ts
import { AbstractDuplexConnection, ConnectionManager } from '../../protocol';
import { AbstractServer, AbstractClient } from '../../lifecycle';
import { DuplexRouter } from '../../types';
import { ConnectionBrokenException, ConnectionRefusedException } from '../../exceptions';

/**
 * Framework-specific pipe translation layer handling raw string frames.
 */
export class WebsocketsDuplexConnection extends AbstractDuplexConnection {
    private ws: any;

    constructor(
        websocket: any,
        router: DuplexRouter,
        clientId?: string,
        manager?: ConnectionManager
    ) {
        super(router, clientId);
        this.ws = websocket;

        if (manager) {
            manager.register(this.clientId, this);
        }

        // Direct network traffic straight down into the handleMessage parser loop
        this.ws.onmessage = async (event: any) => {
            const rawData = typeof event.data !== 'undefined' ? event.data : event;
            await this.handleMessage(rawData.toString());
        };

        this.ws.onclose = async () => {
            if (manager) {
                manager.unregister(this.clientId);
            }
            await this.close();
        };

        this.ws.onerror = async () => {
            if (manager) {
                manager.unregister(this.clientId);
            }
            await this.close();
        };
    }

    protected async _rawSend(payloadStr: string): Promise<void> {
        try {
            // 1 signifies standard WebSocket.OPEN state
            if (this.ws.readyState !== 1) {
                throw new ConnectionBrokenException();
            }
            this.ws.send(payloadStr);
        } catch (err) {
            throw new ConnectionBrokenException(err instanceof Error ? err.message : String(err));
        }
    }

    protected async _rawClose(): Promise<void> {
        if (this.ws.readyState === 0 || this.ws.readyState === 1) { // CONNECTING or OPEN
            this.ws.close();
        }
    }
}

/**
 * Concrete WebSocket Server mapping lifecycle hooks directly to the Node 'ws' package.
 */
export class WebsocketServer extends AbstractServer {
    private port: number;
    private host: string;
    private _server: any | null = null;

    constructor(router: DuplexRouter, host: string, port: number) {
        super(router);
        this.host = host;
        this.port = port;
    }

    protected async _rawStart(): Promise<void> {
        // Dynamic import isolates this backend code from frontend compilation boundaries
        const { WebSocketServer: NodeWSServer } = await import('ws');

        this._server = new NodeWSServer({ host: this.host, port: this.port });

        this._server.on('connection', (websocket: any) => {
            new WebsocketsDuplexConnection(websocket, this.router, undefined, this.manager);
        });

        console.info(`WebSocket Server active on ws://${this.host}:${this.port}`);
    }

    protected async _rawStop(): Promise<void> {
        if (this._server) {
            return new Promise<void>((resolve, reject) => {
                this._server!.close((err: any) => {
                    if (err) reject(err);
                    else {
                        console.info("WebSocket Server shut down cleanly");
                        resolve();
                    }
                });
            });
        }
    }
}

/**
 * Concrete WebSocket Client mapping connection handshakes directly to native web context APIs or ws wrappers.
 */
export class WebsocketClient extends AbstractClient {
    private targetUrl: string;

    constructor(router: DuplexRouter, targetUrl: string) {
        super(router);
        this.targetUrl = targetUrl;
    }

    protected async _createConnection(): Promise<WebsocketsDuplexConnection> {
        return new Promise<WebsocketsDuplexConnection>(async (resolve, reject) => {
            try {
                const WSConstructor = typeof WebSocket !== 'undefined' ? WebSocket : (await import('ws')).default;
                const websocket = new WSConstructor(this.targetUrl);

                websocket.onopen = () => {
                    websocket.onerror = null;
                    resolve(new WebsocketsDuplexConnection(websocket, this.router));
                };

                websocket.onerror = () => {
                    reject(new ConnectionRefusedException());
                };
            } catch (err) {
                reject(new ConnectionRefusedException());
            }
        });
    }
}
